Fix Part of #10700: Refactor LearnerAction Object Factory

DmirtyUsov opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #10700.
2. This PR does the following: refactor **LearnerActionObjectFactory**

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...",
followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] The linter/Karma presubmit checks have passed on my local machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [x] My PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] I have assigned the correct reviewers to this PR (or will leave a
comment with the phrase "@{{reviewer_username}} PTAL" if I don't have
permissions to assign reviewers directly).

## Proof that changes are correct
![image](https://github.com/oppia/oppia/assets/28049705/ffe7f7b5-88cb-43a8-a9b0-ac0169a503eb)

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface. Please also include
videos/screenshots of the developer tools browser console, so that we can be
sure that there are no console errors.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

DmirtyUsov committed:
refactor: migrate LearnerActionObjectFactory to LearnerAction model class

seanlip commented:
@DmirtyUsov Just a note, please make sure you assign reviewers once you're happy with the PR. See the last checkbox in "Essential Checklist" for instructions on how to do that!

DmirtyUsov committed:
Merge branch 'develop' into refactor-learner-action

DmirtyUsov commented:
@brianrodri PTAL
The name ```LearnerAction``` was already taken.
 ```
export type LearnerAction = (...
```
I changed it to
```
export type LearnerActionType = (...
```

seanlip commented:
@DmirtyUsov See my note [here](https://github.com/oppia/oppia/pull/18870#discussion_r1324914745), it might be relevant to your comment here too.

kevintab95 reviewed (approved):
LGTM, thanks @DmirtyUsov!

DmirtyUsov committed:
Merge remote-tracking branch 'upstream/develop' into refactor-learner-action

DmirtyUsov committed:
refactor: change class name LearnerAction to LearnerActionModel

DmirtyUsov commented:
@kevintab95 PTAL

As discussed  https://github.com/oppia/oppia/pull/18870#discussion_r1328112689 with @seanlip  
I changed the name to LearnerActionModel consistency and added enum for Learner Action Types.

seanlip reviewed (approved):
Thanks @DmirtyUsov :) It looks good to me!

DmirtyUsov committed:
Merge branch 'develop' into refactor-learner-action

DmirtyUsov committed:
Merge branch 'develop' into refactor-learner-action

DmirtyUsov committed:
Merge branch 'develop' into refactor-learner-action

DmirtyUsov committed:
Merge branch 'develop' into refactor-learner-action

DmirtyUsov committed:
Merge remote-tracking branch 'upstream/develop' into refactor-learner-action

DmirtyUsov committed:
Merge remote-tracking branch 'upstream/develop' into refactor-learner-action
